proto/message_contents/public_key.proto

```diff
@@ -29,7 +30,8 @@ message UnsignedPublicKey {
 // SignedPublicKey 
 message SignedPublicKey {
     bytes key_bytes = 1;  // embeds an UnsignedPublicKey
-    Signature signature = 2; // signs key_bytes
+    Signature signature = 2; // signs key_bytes, legacy association proof
+    AssociationProof proof = 3; // proves association with a user identity
```

Comment:
@richardhuaaa  definitely hear you on the sequential naming. I'm not opposed to SignedPublicKeyV2 over AssociatedPublicKey.

My comments would be 
1) If we feel like name could be more accurate, then now is the time to fix it -- It will only get more confusing later
2) The goal is developers can will never have to know about SignedPublicKey as it will be deprecated and removed from the code base as the protocol moves to the new flow. 

> SignatureV2
This i think is a conceptual change from the above discussion. Signature is a cryptographic concept, and association proof is a higher level abstraction that contains a Signature. ref: https://github.com/xmtp/proto/pull/39#issuecomment-1440871625

Reply:
I'm a bit confused. We want the concept of `Proof` to be reused beyond just `SignedPublicKeyV2` right? The idea being any time we need a binding we can expect an `AssociationProof` field and run the same verification code. `AssociationProof` or any `Proof` allows us to support self-describing signatures, or non-local signatures, etc and the main goal of this PR is to start baking the concept of Proof into protos.

`SignedPublicKeyV2` can and probably should be the first consumer of whatever `Proof` we come up with, but does not eliminate the need for it. Is that lining up with ya'lls thinking @jazzz @richardhuaaa

Reply:
I for one am finding this discussion very helpful, as its teasing out the parts that need more clarity.

> The idea being any time we need a binding we can expect an AssociationProof field and run the same verification code.

The more we discuss, the fewer conceptual differences I see between a 'AssociationProof' and a SignedPublicKey. An associationProof is a proof that a "key" is associated with blockchainAccount. I'm having a hard time thinking of where else an association is used in the XMTP Protocol outside of keys. 

>SignedPublicKeyV2 can and probably should be the first consumer of whatever Proof we come up with

If we can't find reuse for the AssociationProof, then I'd suggest flattening it for simplicity. Though open.

Reply:
Digging in a little deeper:

https://github.com/xmtp/proto/blob/main/proto/message_contents/public_key.proto#L36-L41

This proto I think summarizes the tension I see.

IdentityKeys are signed by the Wallet, and PreKeys are signed by another Key. They use different inputs to the signature. This manifests in the codebases in interesting ways where the validation of that signature is context dependent. 

I know there has been much circular discussion, but in the end of the day the focus right now is wider support for different wallet signatures.

How do we get to that step the cleanest?

Reply:
I could see that argument too. Basically it boils down to "no more standalone Signature fields, starting with SignedPublicKeyV2". Any time we want a signature to mean something, we create shove the necessary context to accompany the `Signature` into the parent message.

My only worry is that without standardizing that "necessary context" then we might find ourselves in a confusing place. `SignedPublicKeyV2` carries a field with an enum for "SignatureValidation" and "SignatureGeneration". That enum gets duplicated into `SignedXYZ` later down the line. Maybe we just make standalone enums for any options. Any time a `Signature` appears in any proto, it's expected there are `SignatureValidation` and `SignatureGeneration` fields right next to it.

Reply:
^^ but if we do that, then we go full circle back to #39 - why not just put those enums in the Signature proto message?

Reply:
> How do we get to that step the cleanest?

My honest opinion: Do #39 where we make new signature types that carry context in them. We accept that `Signature` in our proto really means `Signature + a bunch of context, aka a Proof e.g. like a JWT json-web-token`.